File: go/lib/linked_list.go
package lib

// Linked List
type LLNode struct {
	Value int
	Next  *LLNode
}

func LL(arr []int) *LLNode {
	var head *LLNode

	for i := len(arr) - 1; i >= 0; i-- {
		value := arr[i]
		head = &LLNode{Value: value, Next: head}
	}

	return head
}

func (head *LLNode) Push(value int) *LLNode {
	return head.LLAddNode(&LLNode{Value: value})
}

func (head *LLNode) LLAddValue(value int) *LLNode {
	return head.Push(value)
}

func (head *LLNode) LLAddNode(node *LLNode) *LLNode {
	if head != nil {
		tail := head
		for tail.Next != nil {
			tail = tail.Next
		}

		tail.Next = node
		return head
	} else {
		return node
	}
}

func (head *LLNode) LLToArray() []int {
	res := []int{}
	curr := head

	for curr != nil {
		res = append(res, curr.Value)
		curr = curr.Next
	}

	return res
}

func (head *LLNode) LLSize() int {
	n := 0

	curr := head
	for curr != nil {
		curr = curr.Next
		n += 1
	}

	return n
}

func (head *LLNode) Any() bool {
	return head != nil
}

func (head *LLNode) Empty() bool {
	return head == nil
}

func (head *LLNode) ToString() string {
	return SliceToString(head.LLToArray())
}

File: ruby/lib/linked_list.rb
# Linked List
class LLNode
  attr_accessor :value, :next_node

  def initialize(value, next_node=nil)
    @value = value
    @next_node = next_node
  end
end

class LinkedList
  attr_reader :head

  def initialize(arr = nil)
    @head = ll(Array(arr))
  end

  def push(value)
    @head = ll_add_value(@head, value)
  end

  def pop
    if @head
      value = @head.value
      @head = @head.next_node
      value
    end
  end

  def to_array
    ll_to_array(@head)
  end

  def empty?
    @head == nil
  end

  def head_value
    @head.value
  end

  def each(&block)
    curr = head
    while curr do
      block.call(curr)
      curr = curr.next_node
    end
  end

  def remove_first
    @head = @head&.next_node
  end
end

def ll(arr)
  head = nil

  arr.reverse.each do |value|
    head = LLNode.new(value, head)
  end

  head
end

def ll_add_value(head, value)
  ll_add_node(head, LLNode.new(value))
end

def ll_add_node(head, node)
  if head
    tail = head
    while tail.next_node do
      tail = tail.next_node
    end

    tail.next_node = node
    head
  else
    node
  end
end

def ll_to_array(head)
  res = []
  curr = head

  while curr do
    res << curr.value
    curr = curr.next_node
  end

  res
end

def ll_size(head)
  n = 0
  curr = head
  while curr do
    curr = curr.next_node
    n += 1
  end
  n
end
